File: 208 Implement Trie (Prefix Tree)/solution.cpp
// Implement a trie with insert, search, and startsWith methods.

// Note:
// You may assume that all inputs are consist of lowercase letters a-z.
#include <unordered_map>
#include <string>
using namespace std;

class TrieNode {
    char val;
    bool isWord;
    unordered_map<char, TrieNode*> children;

public:
    TrieNode(): isWord(false) {}
    TrieNode(char c): val(c), isWord(false) {}
    bool hasChild(const char c) {
        return this->children.count(c) > 0;
    }
    void addChild(const char c) {
        this->children[c] = new TrieNode(c);
    }
    TrieNode* getChild(const char c) {
        return this->children[c];
    }
    void setWord() {
        this->isWord = true;
    }
    bool ifWord() {
        return this->isWord;
    }
};

class Trie {
    TrieNode* root;
public:
    /** Initialize your data structure here. */
    Trie(): root(new TrieNode()) {
        
    }
    
    /** Inserts a word into the trie. */
    void insert(string word) {
        auto track = this->root;
        for (unsigned i = 0; i < word.length(); i++) {
            if (!track->hasChild(word[i])) {
                track->addChild(word[i]);
            }
            track = track->getChild(word[i]);
        }
        track->setWord();
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) {
        auto track = this->root;
        for (unsigned i = 0; i < word.length(); i++) {
            if (!track->hasChild(word[i]))
                return false;
            track = track->getChild(word[i]);
        }
        return track->ifWord();
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) {
        auto track = this->root;
        for (unsigned i = 0; i < prefix.length(); i++) {
            if (!track->hasChild(prefix[i]))    
                return false;
            track = track->getChild(prefix[i]);
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * bool param_2 = obj.search(word);
 * bool param_3 = obj.startsWith(prefix);
 */

File: 208 Implement Trie (Prefix Tree)/solution.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Implement a trie with insert, search, and startsWith methods.

# Note:
# You may assume that all inputs are consist of lowercase letters a-z.

class TrieNode(object):

    def __init__(self, char=''):
        self.isWord = False
        self.children = {}
        self.val = char

class Trie(object):

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        i, track = 0, self.root
        while i < len(word) and word[i] in track.children:
            i, track = i+1, track.children[word[i]]
        while i < len(word):
            track.children[word[i]] = TrieNode(word[i])
            track = track.children[word[i]]
            i += 1
        track.isWord = True

    def search(self, word):
        i, track = 0, self.root
        while i < len(word) and word[i] in track.children:
            i, track = i+1, track.children[word[i]]
        if i == len(word) and track.isWord:
            return True
        return False
        

    def startsWith(self, prefix):
        i, track = 0, self.root
        while i < len(prefix) and prefix[i] in track.children:
            i, track = i+1, track.children[prefix[i]]
        if i == len(prefix):
            return True
        return False
        


if __name__ == "__main__":
    s = Trie()
    s.insert("Trie")
    s.insert("startsWith")
    print(s.search("a"))
    print(s.startsWith("a"))

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)